### PR TITLE
Progressive probing for downloads testing

### PR DESCRIPTION
**Pull request by rotosonic**
Why: downloads was over estimating bandwidth rates
How: complete longer probing to get bandwidth estimate and set number of concurrent connections based on this value
Test: run speedTest locally connected with ethernet device if possible and verify bandwidth results.

### CODE REVIEW

File: public/examples/download/downloadApp.js [hunk 1]
```diff
@@ -161,8 +167,8 @@
                 resultsEl[i].innerHTML = '';
             }
         }
-        downloadProbe();
```
Comment (rotosonic) [inline public/examples/download/downloadApp.js]: remove line downloadProbe()

Comment (rotosonic) [inline public/examples/download/downloadApp.js]: and function

File: public/examples/download/downloadApp.js [hunk 2]
```diff
@@ -161,8 +167,8 @@
...
-
+        //downloadProbe();
+      void (!(testPlan.hasIPv6 === 'IPv6') && setTimeout(function () { !firstRun && downloadTest(testPlan.hasIPv6 ? 'IPv6' : 'IPv4'); }, 500));
```
Comment (rotosonic) [inline public/examples/download/downloadApp.js]: verify and code as boolean

File: public/lib/downloadHttpConcurrentProgress.js [hunk 3]
```diff
@@ -82,19 +93,11 @@
      * onAbort method
      * @return abort object
      */
-    downloadHttpConcurrentProgress.prototype.onTestAbort = function () {
-        if(this._running) {
```
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: for running state the monitor functions controls the end of the test

File: public/lib/downloadHttpConcurrentProgress.js [hunk 4]
```diff
@@ -126,18 +129,26 @@
         }
         this._collectMovingAverages = false;
         //pushing results to an array
-        this._results.push(result);
+        this._results.push(result.bandwidth);
```
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: verify array

File: public/lib/downloadHttpConcurrentProgress.js [hunk 5]
```diff
@@ -126,18 +129,26 @@
...
         //reset Active Tests array
         this._activeTests.length =0;
         //checking if we can continue with the test
         if ((Date.now() - this._beginTime) < this.testLength) {
-            this.start();
+          if(this.isProbing) {
+            this.abortAll();
+            this.probeTotalTime = this.probeTotalTime + result.time;
+            this.probeTotalBytes = this.probeTotalBytes + result.loaded;
+            this.size = (this.probeTimeTimeout-result.time) * result.loaded/result.time;
```
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: check of result.time not zero

File: public/lib/downloadHttpConcurrentProgress.js [hunk 6]
```diff
@@ -205,6 +210,7 @@
                 }
                 singleMovingAverage = singleMovingAverage / lastElem;
                 totalMovingAverage = totalMovingAverage + singleMovingAverage;
+              this._results.push(totalMovingAverage);
```
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: verify using results array

File: public/lib/downloadHttpConcurrentProgress.js (hunk 4 above)
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: remove end of test check from testComplete function

File: public/lib/downloadHttpConcurrentProgress.js [hunk 7]
```diff
@@ -161,15 +172,9 @@
...
-            clearInterval(this.interval);
-            this.abortAll();
-            if (this.finalResults && this.finalResults.length) {
-                this.clientCallbackComplete(this.finalResults);
-            } else {
-                this.clientCallbackError('no measurements obtained');
-            }
-            this._running = false;
+        if(this.isProbing){
+          this.probeTotalTime = this.probeTotalTime + result.totalTime;
```
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: verify not using probeTotalTime

File: public/lib/downloadHttpConcurrentProgress.js [hunk 8]
```diff
@@ -269,7 +268,35 @@
...
         }
-      this.abortAll();
+        this.abortAll();
+      }
+      //check for end of probing
+      if ((Date.now() - this._beginTime) > (this.probeTimeTimeout) && this.isProbing) {
+        this.isProbing = false;
+        this.abortAll();
+        //TODO check on better way to get testing size
+        this.size = ((this.testLength - this.probeTimeTimeout) * this.probeTotalBytes) / (3000 * this.concurrentRuns);
```
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: remove hardcoded probetimeout

File: public/lib/downloadHttpConcurrentProgress.js (hunk 5 above)
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: Due not lower size during actual testing check new size after group run is not lower

File: public/lib/xmlhttprequest.js [hunk 9]
```diff
@@ -206,12 +208,17 @@
...
-          if(isFinite(result.bandwidth)){
-            this.callbackProgress(result);
-            this.prevTime = this.currentTime;
-            this.prevLoad = response.loaded;
+          if (result.totalTime > this.progressIntervalDownload) {
+            var transferDurationSeconds = result.totalTime / 1000;
+            result.bandwidth = transferSizeMbs / transferDurationSeconds;
+            result.loaded = response.loaded;
+            result.startTime = this.startTime;
+            result.chunckLoadedMb = transferSizeMbs;
```
Comment (rotosonic) [inline public/lib/xmlhttprequest.js]: check if still using

File: public/speed-testJS.js [hunk 10]
```diff
@@ -194,8 +200,10 @@
             }
         }
         
-        setTimeout(downloadProbe(),500);
-        //update button text to communicate current state of test as In Progress
+
+      void (!(testPlan.hasIPv6 === 'IPv6') && setTimeout(function () { !firstRun && downloadTest(testPlan.hasIPv6 ? 'IPv6' : 'IPv4'); }, 500));
```
Comment (rotosonic) [inline public/speed-testJS.js]: check ipv6 is boolean

File: public/examples/download/downloadApp.js (hunk 1 above)
Comment (rotosonic) [inline public/examples/download/downloadApp.js]: completed

File: public/lib/downloadHttpConcurrentProgress.js (hunk 4 above)
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: removed array

Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: removed

File: public/lib/downloadHttpConcurrentProgress.js (hunk 5 above)
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: added

Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: added

File: public/lib/downloadHttpConcurrentProgress.js (hunk 7 above)
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: removed

File: public/lib/downloadHttpConcurrentProgress.js (hunk 6 above)
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: removed

File: public/lib/downloadHttpConcurrentProgress.js (hunk 8 above)
Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: removed

**Comment by kemarc** [comment]
+1

File: public/lib/downloadHttpConcurrentProgress.js [hunk 11]
```diff
@@ -125,32 +127,35 @@
...
+                this.probeTotalBytes = this.probeTotalBytes + result.loaded;
+                this.size = (this.probeTimeTimeout - result.time) * result.loaded / result.time;
+              }
+            }catch(error){// jshint ignore:line
+
             }
+
+          }
+          else{
+            if((this.timeout * result.loaded/result.time)< this.size) {
```
Comment (awangc) [inline public/lib/downloadHttpConcurrentProgress.js]: This should be `if (this.size < (this.timeout * result.loaded/result.time))` ?

Comment (rotosonic) [inline public/lib/downloadHttpConcurrentProgress.js]: updated

File: public/lib/downloadHttpConcurrentProgress.js [hunk 12]
```diff
@@ -125,32 +127,37 @@
...
-                if (this.finalResults && this.finalResults.length) {
-                    this.clientCallbackComplete(this.finalResults);
-                } else {
-                    this.clientCallbackError('no measurements obtained');
+          if(this.isProbing) {
+            this.abortAll();
+            try{
+              if(result.time>0) {
+                this.probeTotalBytes = this.probeTotalBytes + result.loaded;
+                if((this.timeout * result.loaded/result.time)> this.size) {
```
Comment (awangc) [inline public/lib/downloadHttpConcurrentProgress.js]: This check and line 142 do not match.

**Comment by awangc** [comment]
Once that last comment is addressed, it has my :+1:

**Pull request merged**